src/Razor/Microsoft.AspNetCore.Razor.Tools/src/RewriteCssCommand.cs

```diff
@@ -162,13 +167,28 @@ protected override void VisitSelector(Selector selector)
                 // Also remove the deep combinator if we matched one
                 if (firstDeepCombinator != null)
                 {
-                    Edits.Add(new DeleteContentEdit { Position = firstDeepCombinator.Start, DeleteLength = firstDeepCombinator.Length });
+                    Edits.Add(new DeleteContentEdit { Position = firstDeepCombinator.Start, DeleteLength = DeepCombinatorText.Length });
                 }
             }
 
-            private static bool IsDeepCombinator(string simpleSelectorText)
+            private int FindPositionBeforeTrailingCombinator(SimpleSelector lastSimpleSelector)
             {
-                return string.Equals(simpleSelectorText, "::deep", StringComparison.Ordinal);
+                // For a selector like "a > ::deep b", the parser splits it as "a >", "::deep", "b".
+                // The place we want to insert the scope is right after "a", hence we need to detect
+                // if the simple selector ends with " >" or similar, and if so, insert before that.
+                var text = lastSimpleSelector.Text;
+                var lastChar = text.Length > 0 ? text[^1] : default;
+                switch (lastChar)
+                {
+                    case '>':
+                    case '+':
+                    case '~':
+                        var trailingCombinatorMatch = TrailingCombinatorRegex.Match(text);
+                        var trailingCombinatorLength = trailingCombinatorMatch.Length;
```

Comment:
Does this need to check if the match succeeded?

Reply:
I considered it but we already know it must succeed given the contents of the regex. Again I don’t mind adding a check but it might be confusing to any developer who’s reasoning about what this code does.

Reply:
Added the check with a comment to be super clear.